File: android/src/main/java/com/margelo/nitro/nitroota/utils/PatchUtils.kt
package com.margelo.nitro.nitroota.utils

import java.io.File
import java.io.IOException
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.MessageDigest

/**
 * Applies NITROBSD patches (bsdiff algorithm, raw blocks) to rebuild a bundle file.
 */
object PatchUtils {
    private val MAGIC = "NITROBSD".toByteArray(Charsets.US_ASCII)
    private const val HEADER_SIZE = 84
    private const val MAX_NEW_SIZE = 256 * 1024 * 1024

    fun sha256(bytes: ByteArray): ByteArray = MessageDigest.getInstance("SHA-256").digest(bytes)

    fun apply(base: File, patch: File, output: File) {
        output.writeBytes(apply(base.readBytes(), patch.readBytes()))
    }

    fun apply(old: ByteArray, patch: ByteArray): ByteArray {
        if (patch.size < HEADER_SIZE || !patch.copyOfRange(0, 8).contentEquals(MAGIC)) {
            throw IOException("Invalid patch header")
        }
        val buffer = ByteBuffer.wrap(patch).order(ByteOrder.LITTLE_ENDIAN)
        val ctrlLen = buffer.getInt(8)
        val diffLen = buffer.getInt(12)
        val newSize = buffer.getInt(16)
        if (ctrlLen < 0 || ctrlLen % 12 != 0 || diffLen < 0 || newSize < 0 || newSize > MAX_NEW_SIZE ||
            HEADER_SIZE.toLong() + ctrlLen + diffLen > patch.size
        ) {
            throw IOException("Corrupt patch header")
        }
        if (!sha256(old).contentEquals(patch.copyOfRange(20, 52))) {
            throw IOException("Base bundle mismatch")
        }

        val out = ByteArray(newSize)
        var ctrl = HEADER_SIZE
        val ctrlEnd = HEADER_SIZE + ctrlLen
        var diff = ctrlEnd
        val diffEnd = ctrlEnd + diffLen
        var extra = diffEnd
        var oldPos = 0L
        var newPos = 0

        while (newPos < newSize) {
            if (ctrl + 12 > ctrlEnd) throw IOException("Corrupt patch control block")
            val x = buffer.getInt(ctrl)
            val y = buffer.getInt(ctrl + 4)
            val z = buffer.getInt(ctrl + 8)
            ctrl += 12
            if (x < 0 || y < 0 || oldPos < 0 || oldPos + x > old.size ||
                newPos.toLong() + x + y > newSize || diff.toLong() + x > diffEnd || extra.toLong() + y > patch.size
            ) {
                throw IOException("Corrupt patch data")
            }
            val o = oldPos.toInt()
            for (i in 0 until x) {
                out[newPos + i] = (old[o + i] + patch[diff + i]).toByte()
            }
            newPos += x
            diff += x
            oldPos += x
            System.arraycopy(patch, extra, out, newPos, y)
            newPos += y
            extra += y
            oldPos += z
        }

        if (!sha256(out).contentEquals(patch.copyOfRange(52, 84))) {
            throw IOException("Patched bundle hash mismatch")
        }
        return out
    }
}

File: ios/PatchUtils.swift
//
//  PatchUtils.swift
//  NitroOta
//
//  Applies NITROBSD patches (bsdiff algorithm, raw blocks) to rebuild a bundle file.
//

import CryptoKit
import Foundation

enum PatchUtils {
    private static let magic = Array("NITROBSD".utf8)
    private static let headerSize = 84
    private static let maxNewSize = 256 * 1024 * 1024

    static func sha256(_ bytes: [UInt8]) -> [UInt8] {
        return Array(SHA256.hash(data: bytes))
    }

    static func apply(base: URL, patch: URL, output: URL) throws {
        let old = [UInt8](try Data(contentsOf: base))
        let patchBytes = [UInt8](try Data(contentsOf: patch))
        let new = try apply(old: old, patch: patchBytes)
        try Data(new).write(to: output, options: .atomic)
    }

    static func apply(old: [UInt8], patch p: [UInt8]) throws -> [UInt8] {
        guard p.count >= headerSize, Array(p[0..<8]) == magic else {
            throw error("Invalid patch header")
        }
        let ctrlLen = Int(readInt32(p, 8))
        let diffLen = Int(readInt32(p, 12))
        let newSize = Int(readInt32(p, 16))
        guard ctrlLen >= 0, ctrlLen % 12 == 0, diffLen >= 0,
              newSize >= 0, newSize <= maxNewSize,
              headerSize + ctrlLen + diffLen <= p.count else {
            throw error("Corrupt patch header")
        }
        guard sha256(old) == Array(p[20..<52]) else {
            throw error("Base bundle mismatch")
        }

        var out = [UInt8](repeating: 0, count: newSize)
        var ctrl = headerSize
        let ctrlEnd = headerSize + ctrlLen
        var diff = ctrlEnd
        let diffEnd = ctrlEnd + diffLen
        var extra = diffEnd
        var oldPos = 0
        var newPos = 0

        while newPos < newSize {
            guard ctrl + 12 <= ctrlEnd else { throw error("Corrupt patch control block") }
            let x = Int(readInt32(p, ctrl))
            let y = Int(readInt32(p, ctrl + 4))
            let z = Int(readInt32(p, ctrl + 8))
            ctrl += 12
            guard x >= 0, y >= 0, oldPos >= 0, oldPos + x <= old.count,
                  newPos + x + y <= newSize, diff + x <= diffEnd, extra + y <= p.count else {
                throw error("Corrupt patch data")
            }
            for i in 0..<x {
                out[newPos + i] = old[oldPos + i] &+ p[diff + i]
            }
            newPos += x
            diff += x
            oldPos += x
            for i in 0..<y {
                out[newPos + i] = p[extra + i]
            }
            newPos += y
            extra += y
            oldPos += z
        }

        guard sha256(out) == Array(p[52..<84]) else {
            throw error("Patched bundle hash mismatch")
        }
        return out
    }

    private static func readInt32(_ b: [UInt8], _ i: Int) -> Int32 {
        let value = UInt32(b[i]) | UInt32(b[i + 1]) << 8 | UInt32(b[i + 2]) << 16 | UInt32(b[i + 3]) << 24
        return Int32(bitPattern: value)
    }

    private static func error(_ message: String) -> NSError {
        return NSError(domain: "OtaManager", code: -1, userInfo: [NSLocalizedDescriptionKey: message])
    }
}
